compiler/optimizer/OMRSimplifierHandlers.cpp

```diff
@@ -12801,6 +12801,196 @@ TR::Node *su2dSimplifier(TR::Node * node, TR::Block * block, TR::Simplifier * s)
    return node;
    }
 
+
+/**
+ * \brief
+ *
+ * Remove add/subtract operations under comparison nodes to save an
+ * addition or subtraction instruction during codegen. This simplification
+ * applies to:
+ *
+ * 1. Signed comparisons of byte, short, int, and long values.
+ * 2. Unsigned (in)equality of byte, short, int, and long values.
+ *
+ * \details
+ *
+ * \verbatim
+ * As an example, assume the following is the original expression
+ *        x + oldConst1 > oldConst2
+ *
+ * This expression can be simplified to
+ *        x > newConst
+ *        where newConst = oldConst2 - oldConst1
+ * The newConst calculation should not produce overflow or underflow.
+ * \endverbatim
+ *
+ * Also note that for unsigned comparison types other than equality and inequality,
+ * add/sub simplifications are not possible due to how values are interpreted
+ * in the current IL nodes. Essentially, the value of an iconst node is neither
+ * signed nor unsigned; and its parent is free to interpret its value.
+ *
+ * For example, the following transformation is incorrect:
+ *
+ * \varbatim
+ *
+ * Before:
+ *
+ * ifiucmplt --> goto ...
+ *     isub (cannotOverflow)
+ *         iload x
+ *         iconst -1
+ *     iconst 100
+ *
+ * After:
+ *
+ * ifiucmplt --> goto ...
+ *     iload x
+ *     iconst 99
+ *
+ * \endverbatim
+ *
+ * If x were -1, the first tree would evaluate to true since it's comparing if zero is less than 100.
+ * But in the second tree, ifiucmplt interprets the value of x as UINT_MAX. Thus,
+ * the second tree is comparing UINT_MAX with 99; and the whole comparison node evalutes
+ * to false.
+ *
+ * Unsigned equality and inequality comparisons are not sensitive to sign interpretation problems.
+ *
+ * \param node          the comparison node being simpilfied
+ * \param s             the simplifier object
+*/
+TR::Node* removeArithmeticsUnderIntegralCompare(TR::Node* node,
+                                                TR::Simplifier * s)
+   {
+   if (s->comp()->getOption(TR_DisableIntegerCompareSimplification) ||
+         !(node->getOpCode().isBooleanCompare()   // The node must be an integer comparison node.
+            && node->getNumChildren() > 0
+            && node->getFirstChild()->getOpCode().isInteger()))
+      return node;
+
+   TR::Node* opNode      = node->getFirstChild();
+   TR::Node* secondChild = node->getSecondChild();
+   TR::ILOpCodes op      = opNode->getOpCodeValue();
+
+   // Get recognized operation types. Unsigned IL OpCodes are deprecated and not supported here.
+   bool isAddOp = (op == TR::iadd || op == TR::ladd || op == TR::sadd || op == TR::badd);
+   bool isSubOp = (op == TR::isub || op == TR::lsub || op == TR::ssub || op == TR::bsub);
+
+   bool isUnsignedCompare = node->getOpCode().isUnsignedCompare();
+   bool isEqOrNotEqCompare = node->getOpCode().isCompareForEquality();
+
+   bool isSignednessTypeSupported = isEqOrNotEqCompare || (!isUnsignedCompare && opNode->chkCannotOverflow());
```

Comment:
Should this be using the "cannotOverflow" query ? I believe that is what the rest of the optimizer uses.

Reply:
> Should this be using the "cannotOverflow" query ? I believe that is what the rest of the optimizer uses.

Should `chkCannotOverflow` exist then? What is it's purpose? Is this duplication we can remove?

Reply:
I believe the chk* variants of each query are called from the print* routines that exist for each Node query (to print Node flags in a log file). The chk* variants of the queries check the opcode and the flags value (both have to be checked before a particular flags value string can be printed in the log file) whereas the queries themselves check the flags value only (on the assumption that the caller is already aware that the query being done is relevant for the particular opcode).

Reply:
Changed the query to use `cannotOverflow()`.